src/docfx/validation/CustomRuleProvider.cs

```diff
@@ -0,0 +1,255 @@
+            {
+                error = WithCustomRule(error, customRule);
+            }
+            return error;
+        }
+
+        public static Error WithCustomRule(Error error, CustomRule customRule, bool? enabled = null)
+        {
+            var level = customRule.Severity ?? error.Level;
+
+            if (level != ErrorLevel.Off && customRule.ExcludeMatches(error.OriginalPath ?? error.Source?.File?.Path ?? ""))
+            {
+                level = ErrorLevel.Off;
+            }
+
+            if (enabled != null && !enabled.Value)
+            {
+                level = ErrorLevel.Off;
+            }
+
+            var message = error.Message;
+
+            if (!string.IsNullOrEmpty(customRule.Message))
+            {
+                try
+                {
+                    message = string.Format(customRule.Message, error.MessageArguments);
+                }
+                catch (FormatException)
+                {
+                    message += "ERROR: custom message format is invalid, e.g., too many parameters {n}.";
+                }
+            }
+
+            message = string.IsNullOrEmpty(customRule.AdditionalMessage) ?
+                message : $"{message}{(message.EndsWith('.') ? "" : ".")} {customRule.AdditionalMessage}";
+
+            return new Error(
+                level,
+                string.IsNullOrEmpty(customRule.Code) ? error.Code : customRule.Code,
+                message,
+                error.MessageArguments,
+                error.Source,
+                error.PropertyPath,
+                error.OriginalPath,
+                customRule.PullRequestOnly);
+        }
+
+        private bool TryGetCustomRule(
+            Error error,
+            [MaybeNullWhen(false)] out CustomRule customRule)
+        {
+            if (_customRules.TryGetValue(error.Code, out var customRules))
+            {
+                foreach (var rule in customRules)
+                {
+                    if (rule.PropertyPath != null)
+                    {
+                        // compare with code + propertyPath + contentType
+                        var source = error.Source?.File;
+                        var pageType = source != null ? _documentProvider.GetPageType(source) : null;
+                        if (rule.PropertyPath.Equals(error.PropertyPath, StringComparison.Ordinal) && rule.ContentTypes.Contains(pageType))
+                        {
+                            customRule = rule;
+                            return true;
+                        }
+                        continue;
+                    }
+                    else
+                    {
+                        customRule = rule; // system error
+                        return true;
+                    }
+                }
+            }
+            customRule = null;
+            return false;
+        }
+
+        private Dictionary<string, List<CustomRule>> MergeCustomRules(
```

Comment:
so in the future, can we merge `MetadataCustomeRules`, `ContentValidationRules` and `BuildValidationRules` together into configuration?

Reply:
No. For `ContentValidationRules`, the configuration just includes rules which pr-only=true but includes all of `BuildValidationRules`.